### ISSUE TITLE
Potential issue in categorical critic

### ISSUE THREAD

**Issue by I-AM-PROTO** [younggyoseo/FastTD3#29]
Hello FastTD3 authors,
I think I found an issue in the categorical critic, specifically in the `projection` function.

https://github.com/younggyoseo/FastTD3/blob/2d83b6a7f6560c38c1cc9972ef59b50f1db43e9b/fast_td3/fast_td3_simbav2.py#L286-L295

From what I can understand, `l_mask` and `u_mask` handles an edge case where any value of `target_z` lands at the exact edge between the bins, which can happen for e.g., reward 0 samples.

We can solve this by pushing `l` down by one (unless it's the lowest bin), but the current implementation is actually pushing _both_ `l` and `u` down and up at the same time.

As a result, some projected categorical distributions don't sum to 1 but something larger:

<img width="636" height="111" alt="Image" src="https://github.com/user-attachments/assets/280fbefc-81ef-4a74-8a52-0a78c37a9351" />

Indeed, the sample (idx 17) has reward 0, with `l_mask` and `u_mask` both flagged true at the same position.

<img width="629" height="395" alt="Image" src="https://github.com/user-attachments/assets/e7f35ec3-099e-4c9b-90a2-3a71dca9579d" />

To fix this, I suggest to simply revise the code to:

```
b = (target_z - self.v_min) / delta_z 
l = torch.floor(b).long() 
u = torch.ceil(b).long() 
  
is_int = (l == u)
l = torch.where(is_int & (l > 0), l - 1, l)
u = torch.where(is_int & (l == 0), u + 1, u)
```

Empirically, I found that implementing this fix leads to a significant drop in the critic loss (in Mujoco Playground), which now reaches near 0 instead of the previous 2.1-2.2. Though, I am unsure whether it will affect the final performance of FastTD3 (or +SimbaV2 version).

**Comment by younggyoseo** [comment]
Hi Donghu, this is a good catch, I appreciate it! But I'm a bit surprised -- is the run, with loss close to 0, properly converging to a good gait? I'm asking because it's typically super difficult to get near-zero loss with cross-entropy loss, especially considering that we're using target Q-values that constantly changes the target distribution.

Also with regard to the code -- I think your changes make sense, but one thing is, when updating `u`, `l` was already changed in the previous line, so there's one case where `l` goes down to 0 but again goes up to 1. I would prefer keeping the original logic of introducing lower/upper masks to avoid this. For instance,

```
lower_mask = is_int & (l > 0)
upper_mask = is_int & (l == 0)

lower = ...
upper = ...
```

I would appreciate if you could make the PR for this -- thanks!

**Comment by younggyoseo** [comment]
^ another data point: I applied the fix to the other codebase that was using the same projection code, and couldn't observe a significant difference in learning curves (observed that maximum Q-values observed throughout training have gone down, as expected)

**Comment by I-AM-PROTO** [comment]
I mistook the value loss being near 0, sorry about that. To be more accurate, `qf_loss` dropped below previous convergence point (2.2) after implementing the fix, although it eventually converged to the same level.

<img width="500" height="230" alt="Image" src="https://github.com/user-attachments/assets/6a29e937-d50b-48c1-8a6e-29b3e9553fb3" />

I ran a quick experiment on `T1JoystickFlatTerrain` and FastTD3+SimbaV2 (with 3 seeds). I think the curve did change; even though the highest score stayed the same the 'degradation' did not occur. I think this was also observed in #26 as well?

<img width="500" height="230" alt="Image" src="https://github.com/user-attachments/assets/f9c059d4-1975-46aa-aec3-52692df5d688" />

<img width="500" height="230" alt="Image" src="https://github.com/user-attachments/assets/bd7a3469-b8c7-4a44-be29-49b810c77e2a" />

Funnily enough the gait also changed; the gait distance got heavily shortened especially (although it seems that the upper body became less shaky?).

Before: https://imgur.com/a/RppgcUf
After: https://imgur.com/a/neZlIsK

**Comment by younggyoseo** [comment]
Thank you for additional update! I also merged the PR, amazing!
